language/stdlib/modules/association.move

```diff
@@ -109,6 +114,156 @@ module Association {
     acquires PrivilegedCapability {
         Transaction::assert(addr_is_association(addr), 1002);
     }
-}
 
+// **************** SPECIFICATIONS ****************
+// Note: I experimented with using Markdown to organize this, but I haven't converted
+// it to work with the new doc generator, yet.
+
+// # Specifications for association.move
+
+// This is preliminary.  This is one of the first real module libraries with global
+// specifications, and everything is evolving.
+
+    spec module {
+        pragma verify = true;
+
+        // This mirrors the Move function root_address()
+        define spec_root_address(): address { 0xA550C18 }
+
+        // mirrors Move addr_is_association
+        define spec_addr_is_association(addr: address): bool {
+            exists<PrivilegedCapability<T>>(addr)
+            && global<PrivilegedCapability<T>>(addr).is_certified
+        }
+     }
+
+    // ## Management of privileges
+    // TODO: With grant/remove etc., separately specify that only root may grant or remove privileges
+
+    // ### Management of Root {} marker.
+
+    // The root_address is marked by a Root{} object that is stored at that place only.
+    spec module {
+        // enclose this in a function so we can easily repeat the same property in the invariant
+        // and as a post-condition to initialize (the base case).
+        // SPEC: Only the root address has a Root{} resource.
+        define root_invariant(): bool {
+            all(addresses(), |a| exists<Root>(a) ==> (a == spec_root_address()))
+        }
+    }
+    spec schema OnlyRootAddressHasRootPrivilege {
+        invariant root_invariant();
+    }
+    spec module {
+        // apply to all functions except "initialize" and functions that "initialize" calls
+        // before the invariant is established.
+        // Note: All the called functions OBVIOUSLY cannot affect the invariant.
+        // TODO: Try to find a better approach to this that does not require excepting functions.
+        apply OnlyRootAddressHasRootPrivilege to *<T>, *
```

Comment:
We should open a bug that one only needs to provide one `*` here. I think the rule should be that all type parameters on the left side must be bound by type parameters in the pattern. A schema with no type parameters can be matched against any function.

Reply:
There's just a bug in my spec, not in the Prover, right?